library/spdm_transport_pcidoe_lib/libspdm_doe_pcidoe.c

```diff
@@ -301,6 +301,72 @@ libspdm_return_t libspdm_pci_doe_decode_discovery_request(size_t transport_messa
     return LIBSPDM_STATUS_SUCCESS;
 }
 
+/**
+ * Decode a DOE discovery request message to get the DOE Discovery Version field.
+ * DOE Discovery Version is introduced in PCIE Spec 6.1 Section 6.30.1.1.
+ *
+ * @param  transport_message_size               Size in bytes of the transport message data buffer.
+ * @param  transport_message                    A pointer to a source buffer to store the transport message.
+ * @param  version                              A pointer to a destination to store the DOE Discovery Version.
+ *
+ * @retval LIBSPDM_STATUS_SUCCESS               The message is encoded successfully.
+ * @retval LIBSPDM_STATUS_INVALID_PARAMETER     The message is NULL or the message_size is zero.
+ **/
+libspdm_return_t libspdm_pci_doe_decode_discovery_request_version(size_t transport_message_size,
+                                                                  const void *transport_message,
+                                                                  uint8_t *version)
+{
+    const pci_doe_data_object_header_t *pci_doe_header;
+    uint32_t length;
+    const uint8_t *message;
+
+    LIBSPDM_ASSERT(transport_message_size > sizeof(pci_doe_data_object_header_t));
+    if (transport_message_size <= sizeof(pci_doe_data_object_header_t)) {
+        return LIBSPDM_STATUS_INVALID_MSG_SIZE;
+    }
+
+    pci_doe_header = transport_message;
+    if (pci_doe_header->vendor_id != PCI_DOE_VENDOR_ID_PCISIG) {
+        return LIBSPDM_STATUS_INVALID_MSG_FIELD;
+    }
+
+    switch (pci_doe_header->data_object_type) {
+    case PCI_DOE_DATA_OBJECT_TYPE_DOE_DISCOVERY:
+        /*
+         * Check to see if we received a DOE discovery message.
+         * DOE discovery is not part of the SPDM spec, instead it's part
+         * of the PCIe DOE spec. DOE discovery is mandatory for all
+         * implementations.
+         *
+         * DOE Discovery Version is introduced in PCIE Spec 6.1 Section 6.30.1.1.
+         * It is Byte-1 in DOE discovery request.
+         */
+        message = (const uint8_t *)transport_message + sizeof(pci_doe_data_object_header_t) + 1;
+        if (version != NULL) {
+            *version = *message;
+        }
+        break;
+    default:
+        return LIBSPDM_STATUS_UNSUPPORTED_CAP;
+    }
+
+    if (pci_doe_header->reserved != 0) {
+        return LIBSPDM_STATUS_INVALID_MSG_FIELD;
+    }
+    if (pci_doe_header->length >= PCI_DOE_MAX_SIZE_IN_DW) {
+        return LIBSPDM_STATUS_INVALID_MSG_SIZE;
+    } else if (pci_doe_header->length == 0) {
+        length = PCI_DOE_MAX_SIZE_IN_BYTE;
+    } else {
+        length = pci_doe_header->length * sizeof(uint32_t);
+    }
+    if (length != transport_message_size) {
+        return LIBSPDM_STATUS_INVALID_MSG_SIZE;
+    }
```

Comment:
It seems a bit unnecessary to do all of these checks again. Is it not worth modifying ` libspdm_pci_doe_decode_discovery_request()` to support the `version` field?

Or even just calling ` libspdm_pci_doe_decode_discovery_request()` from `libspdm_pci_doe_decode_discovery_request_version()` so we don't have to duplicate all of these checks

Reply:
> It seems a bit unnecessary to do all of these checks again. Is it not worth modifying ` libspdm_pci_doe_decode_discovery_request()` to support the `version` field?
> 
To support the version field in libspdm_pci_doe_decode_discovery_request() will add a new in/out parameter. That will break the existing code which calls libspdm_pci_doe_decode_discovery_request(). Furthermore, "version" field doesn't exist in DOE Discovery Request before PCIE Spec 6.1.

> Or even just calling ` libspdm_pci_doe_decode_discovery_request()` from `libspdm_pci_doe_decode_discovery_request_version()` so we don't have to duplicate all of these checks
>
I am OK to call libspdm_pci_doe_decode_discovery_request()` from libspdm_pci_doe_decode_discovery_request_version()`